<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Research – Genevera Allen</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Genevera Allen</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="././index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="././research.html"> 
<span class="menu-text">Research</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="././group.html"> 
<span class="menu-text">Group</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="././teaching.html"> 
<span class="menu-text">Teaching</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="././software.html"> 
<span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="././news.html"> 
<span class="menu-text">News</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Research</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="data-driven-discovery" class="level2">
<h2 class="anchored" data-anchor-id="data-driven-discovery">Data-Driven Discovery</h2>
<p><strong>Genevera’s group develops new statistical machine learning tools to help people make reliable discoveries from large and complex data sets, especially in neuroscience and biomedicine.</strong></p>
<p><a href="https://scholar.google.com/citations?hl=en&amp;user=gIUd12QAAAAJ&amp;view_op=list_works&amp;sortby=pubdate"><button type="button" class="btn btn-primary">View Full List of Publications via Google Scholar</button></a></p>
<p><br></p>
</section>
<section id="research-areas" class="level2">
<h2 class="anchored" data-anchor-id="research-areas">Research Areas</h2>
<section id="statistical-machine-learning" class="level3">
<h3 class="anchored" data-anchor-id="statistical-machine-learning">Statistical Machine Learning</h3>
<section id="graphical-models" class="level4">
<h4 class="anchored" data-anchor-id="graphical-models">Graphical Models</h4>
<div class="grid">
<div class="g-col-4">
<p><img src="images/Graphical-models-data-integration.jpg" class="img-fluid"></p>
</div>
<div class="g-col-8">
<p>We develop new types of probabilistic graphical models and graph learning strategies for representing, discovering, and visualizing relationships in large data sets. Our work includes developing new classes of graphical models for diverse data types and multi-modal data as well as new graph learning strategies to tackle challenges encountered in neuroscience and beyond.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>E. Yang, P. Ravikumar, G. I. Allen, and Z. Liu, “Graphical Models via Univariate Exponential Family Distributions”, <em>Journal of Machine Learning Research</em>, <strong>16</strong>:3813-3847, 2015. <a href="https://www.jmlr.org/papers/volume16/yang15a/yang15a.pdf">link</a></li>
<li>E. Yang, Y. Baker, P. Ravikumar, G. I. Allen, and Z. Liu, “Mixed Graphical Models via Exponential Families”, In <em>Artificial Intelligence and Statistics (AISTATS)</em>, oral presentation, 2014. <a href="https://proceedings.mlr.press/v33/yang14a.pdf">link</a></li>
<li>L. Zheng and G. I. Allen, “Graphical Model Inference with Erosely Measured Data”, (To Appear) <em>Journal of American Statistical Association: Theory &amp; Methods</em>, arXiv:2210.11625, 2023. <a href="https://arxiv.org/pdf/2210.11625">link</a></li>
<li>M. Wang and G. I. Allen, “Thresholded Graphical Lasso Adjusts for Latent Variables”, <em>Biometrika</em>, <strong>110</strong>:3, 681-697, 2023. <a href="https://par.nsf.gov/servlets/purl/10470624">link</a></li>
<li>G. Vinci, G. Dasarathy, and G. I. Allen, “Graph Quilting: Graphical Model Selection from Partially Observed Covariances”, arXiv:1912.05573, 2020. <a href="https://arxiv.org/pdf/1912.05573">link</a></li>
<li>E. Yang, P. Ravikumar, G. I. Allen, and Z. Liu, “On Poisson Graphical Models”, In <em>Advances in Neural Information Processing Systems (NeurIPS)</em>, 2013. <a href="https://proceedings.neurips.cc/paper/2013/file/43feaeeecd7b2fe2ae2e26d917b6477d-Paper.pdf">link</a></li>
<li>A. Chang and G. I. Allen, “Subbotin Graphical Models for Extreme Value Dependencies with Applications to Functional Neuronal Connectivity”, <em>Annals of Applied Statistics</em>, <strong>17</strong>:3, 2364-2386, 2023. <a href="https://arxiv.org/pdf/2106.11554">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="data-integration" class="level4">
<h4 class="anchored" data-anchor-id="data-integration">Data Integration</h4>
<div class="grid">
<div class="g-col-6">
<p><img src="images/multi_modal_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-6">
<p>Large data sets are often diverse, with multiple types of features measured on the same set of subjects or observations. We have developed a variety of interpretable machine learning techniques for discovering joint patterns in this so-called mixed multi-modal data.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>T. M. Tang and G. I. Allen, “Integrated Principal Components Analysis”, <em>Journal of Machine Learning Research</em>, <strong>22</strong>:198, 1-71, 2021. <a href="https://www.jmlr.org/papers/volume22/20-084/20-084.pdf">link</a></li>
<li>M. Wang and G. I. Allen, “Integrative Generalized Convex Clustering Optimization and Feature Selection for Mixed Multi-View Data”, <em>Journal of Machine Learning Research</em>, <strong>22</strong>:1-73, 2021. <a href="https://www.jmlr.org/papers/volume22/19-1012/19-1012.pdf">link</a></li>
<li>Y. Baker, T. M. Tang, and G. I. Allen, “Feature Selection for Data Integration with Mixed Multiview Data”, <em>Annals of Applied Statistics</em>, <strong>14</strong>:4, 1676-1698, 2020. <a href="https://arxiv.org/pdf/1903.11232">link</a></li>
<li>J. Liu, L. Zheng, Z. Zhang, and G. I. Allen, “Joint Semi-Symmetric Tensor PCA for Integrating Multi-modal Populations of Networks”, arXiv:2312.14416, 2023. <a href="https://arxiv.org/pdf/2312.14416">link</a></li>
<li>Y. W. Wan, G. I. Allen, and Z. Liu, “TCGA2STAT: Simple TCGA Data Access for Integrated Statistical Analysis in R”, <em>Bioinformatics</em>, <strong>32</strong>:6, 952-954, 2016. <a href="https://scholar.google.com/citations?view_op=view_citation&amp;hl=en&amp;user=gIUd12QAAAAJ&amp;cstart=20&amp;pagesize=80&amp;sortby=pubdate&amp;citation_for_view=gIUd12QAAAAJ:RHpTSmoSYBkC">link</a></li>
<li>E. Yang, Y. Baker, P. Ravikumar, G. I. Allen, and Z. Liu, “Mixed Graphical Models via Exponential Families”, In <em>Artificial Intelligence and Statistics (AISTATS)</em>, oral presentation, 2014. [link](https://proceedings.mlr.press/v33/yang14a.pdf</li>
</ul>
</details>
</div>
</div>
</section>
<section id="clustering" class="level4">
<h4 class="anchored" data-anchor-id="clustering">Clustering</h4>
<div class="grid">
<div class="g-col-3">
<p><img src="images/Clustering-thumbnail.jpg" class="img-fluid"></p>
</div>
<div class="g-col-9">
<p>Clustering seeks to find groups in large data sets. We have developed several convex clustering approaches that offer reliable, principled, and flexible strategies along with built-in visualizations for clustering.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>E. C. Chi, G. I. Allen, and R. Baraniuk, “Convex Biclustering”, <em>Biometrics</em>, <strong>73</strong>:1, 10-19, 2017. <a href="https://citeseerx.ist.psu.edu/document?repid=rep1&amp;type=pdf&amp;doi=8d15ab3ba651eacb805fe7d540f7aa4c6a955ca7">link</a></li>
<li>M. Weylandt, J. Nagorski, and G. I. Allen, “Dynamic Visualization and Fast Computation for Convex Clustering via Algorithmic Regularization”, <em>Journal of Computational and Graphical Statistics</em>, <strong>29</strong>:1, 87-96, 2020. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC7518335/">link</a></li>
<li>M. Wang and G. I. Allen, “Integrative Generalized Convex Clustering Optimization and Feature Selection for Mixed Multi-View Data”, <em>Journal of Machine Learning Research</em>, <strong>22</strong>:1-73, 2021. <a href="https://www.jmlr.org/papers/volume22/19-1012/19-1012.pdf">link</a></li>
<li>Y. Baker, T. M. Tang, and G. I. Allen, “Feature Selection for Data Integration with Mixed Multiview Data”, <em>Annals of Applied Statistics</em>, <strong>14</strong>:4, 1676-1698, 2020. <a href="https://arxiv.org/pdf/1903.11232">link</a></li>
<li>L. Gan and G. I. Allen, “Fast and Interpretable Consensus Clustering via Minipatch Learning”, <em>PLoS Computational Biology</em>, <strong>18</strong>:10, e1010577, 2022. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010577">link</a></li>
<li>M. Wang, T. Yao, and G. I. Allen, “Supervised Convex Clustering”, <em>Biometrics</em>, <strong>79</strong>:4, 3846-3858, 2023. <a href="https://academic.oup.com/biometrics/article/79/4/3846/7587579">link</a></li>
<li>L. Zheng, A. Chang, and G. I. Allen, “Cluster Quilting: Spectral Clustering for Patchwork Learning”, arXiv:2406.13833, 2024. <a href="https://arxiv.org/pdf/2406.13833">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="dimension-reduction" class="level4">
<h4 class="anchored" data-anchor-id="dimension-reduction">Dimension Reduction</h4>
<div class="grid">
<div class="g-col-4">
<p><img src="images/pattern-recognition-or-dimension-reduction-e1632928373812.jpg" class="img-fluid"></p>
</div>
<div class="g-col-8">
<p>Dimension reduction techniques are used for visualizing, exploring, and discovering patterns in large data sets. We have developed many dimension reduction techniques for complex and structured data; these include sparse tensor decompositions and generalizations of PCA for structured or multi-modal data.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>G. I. Allen, L. Grosenick and J. Taylor, “A Generalized Least Squares Matrix Decomposition”, <em>Journal of the American Statistical Association: Theory and Methods</em>, <strong>109</strong>:505, 145-159, 2014. <a href="https://www.researchgate.net/profile/Logan-Grosenick/publication/48208333_A_Generalized_Least-Square_Matrix_Decomposition/links/573b544c08ae298602e45243/A-Generalized-Least-Square-Matrix-Decomposition.pdf">link</a></li>
<li>G. I. Allen, “Sparse Higher-Order Principal Components Analysis”, In <em>Artificial Intelligence and Statistics (AISTATS)</em>, 27-36, 2012. <a href="https://proceedings.mlr.press/v22/allen12/allen12.pdf">link</a></li>
<li>T. M. Tang and G. I. Allen, “Integrated Principal Components Analysis”, <em>Journal of Machine Learning Research</em>, <strong>22</strong>:198, 1-71, 2021. <a href="https://www.jmlr.org/papers/volume22/20-084/20-084.pdf">link</a></li>
<li>G. I. Allen, and M. Weylandt, “Sparse and Functional Principal Components Analysis”, In <em>Proceedings of the IEEE Data Science Workshop</em>, 2019. <a href="https://arxiv.org/pdf/1309.2895">link</a></li>
<li>G. I. Allen and M. Maletic-Savatic, “Sparse Non-negative Generalized PCA with Applications to Metabolomics”, <em>Bioinformatics</em>, <strong>27</strong>:21, 3029-3035, 2011. <a href="https://academic.oup.com/bioinformatics/article/27/21/3029/218921">link</a></li>
<li>J. Liu, L. Zheng, Z. Zhang, and G. I. Allen, “Joint Semi-Symmetric Tensor PCA for Integrating Multi-modal Populations of Networks”, arXiv:2312.14416, 2023. <a href="https://arxiv.org/pdf/2312.14416">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="sparsity-high-dimensional-statistics" class="level4">
<h4 class="anchored" data-anchor-id="sparsity-high-dimensional-statistics">Sparsity &amp; High-Dimensional Statistics</h4>
<div class="grid">
<div class="g-col-4">
<p><img src="images/Regularization-paths-structured-sparsity.jpg" class="img-fluid"></p>
</div>
<div class="g-col-8">
<p>Much of our research lies in the area of high-dimensional statistics, where the number of features exceeds the number of samples. A major focus in this area is on sparsity and structured sparsity to enhance feature selection and interpretability.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>F. Campbell and G. I. Allen, “Within Group Variable Selection through the Exclusive Lasso”, <em>Electronic Journal of Statistics</em>, <strong>11</strong>:2, 4220-4257, 2017. <a href="https://projecteuclid.org/journals/electronic-journal-of-statistics/volume-11/issue-2/Within-group-variable-selection-through-the-Exclusive-Lasso/10.1214/17-EJS1317.full">link</a></li>
<li>G. I. Allen, “Automatic Feature Extraction via Weighted Kernels and Regularization”, <em>Journal of Computational and Graphical Statistics</em>, <strong>22</strong>:2, 284-299, 2013. <a href="https://www.tandfonline.com/doi/abs/10.1080/10618600.2012.681213">link</a></li>
<li>Y. Baker, T. M. Tang, and G. I. Allen, “Feature Selection for Data Integration with Mixed Multiview Data”, <em>Annals of Applied Statistics</em>, <strong>14</strong>:4, 1676-1698, 2020. <a href="https://arxiv.org/pdf/1903.11232">link</a></li>
<li>A. Chang, M. Wang, and G. I. Allen, “Sparse Regression for Extreme Values”, <em>Electronic Journal of Statistics</em>, <strong>15</strong>:2, 5995-6035, 2021. <a href="https://par.nsf.gov/servlets/purl/10470614">link</a></li>
<li>Y. Hu, E. C. Chi, and G. I. Allen, “ADMM Algorithmic Regularization Paths for Sparse Statistical Learning”, In <em>Splitting Methods in Communication and Imaging, Science and Engineering</em>, R. Glowinski, W. Yin, and S. Osher (eds), Springer, 433-459, 2016. <a href="https://arxiv.org/pdf/1504.06637">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="tensors" class="level4">
<h4 class="anchored" data-anchor-id="tensors">Tensors</h4>
<div class="grid">
<div class="g-col-6">
<p><img src="images/tensor_cp_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-6">
<p>Directly working with tensor, or multi-way array, data yields many computational and statistical advantages. Our work has focused on developing interpretable tensor decomposition strategies with applications in neuroscience, chemometrics, and genomics.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>G. I. Allen, “Sparse Higher-Order Principal Components Analysis”, In <em>Artificial Intelligence and Statistics (AISTATS)</em>, 27-36, 2012. <a href="https://proceedings.mlr.press/v22/allen12/allen12.pdf">link</a></li>
<li>G. I. Allen, “Multi-way Functional Principal Components Analysis”, In <em>IEEE International Workshop on Computational Advances in Multi-Sensor Adaptive Processing</em>, 2013. <a href="https://signalprocessingsociety.org/CAMSAP2013/papers/1569807425.pdf">link</a></li>
<li>Z. Zhang, G. I. Allen, H. Zhu, D. Dunson, “Tensor Network Factorizations: Relationships between Human Brain Structural Connectomes and Traits”, <em>NeuroImage</em>, <strong>197</strong>:330-343, 2019. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6613218/">link</a></li>
<li>J. Liu, L. Zheng, Z. Zhang, and G. I. Allen, “Joint Semi-Symmetric Tensor PCA for Integrating Multi-modal Populations of Networks”, arXiv:2312.14416, 2023. <a href="https://arxiv.org/pdf/2312.14416">link</a></li>
<li>K. Geyer, F. Campbell, A. Chang, J. Magnotti, M. Beauchamp, G. I. Allen, “Interpretable Visualization and Higher-Order Dimension Reduction for ECoG Data”, In <em>Proceedings of the The International Workshop on Big Data Reduction</em> held with the <em>2020 IEEE International Conference on Big Data</em>, 2020. <a href="https://arxiv.org/pdf/2011.09447">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="ensemble-learning" class="level4">
<h4 class="anchored" data-anchor-id="ensemble-learning">Ensemble Learning</h4>
<div class="grid">
<div class="g-col-3">
<p><img src="images/fig_minipatch_illustration.png" class="img-fluid"></p>
</div>
<div class="g-col-9">
<p>Recently, we have begun developing new computationally efficient ensemble learning strategies called minipatch learning that also lead to improved statistical properties and interpretability.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>T. Yao, D. LeJeune, H. Javadi, R. G. Baraniuk, and G. I. Allen, “Minipatch Learning as Implicit Ridge-Like Regularization”, In <em>IEEE International Conference on Big Data and Smart Computing (BigComp)</em>, 2021. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC8570556/">link</a></li>
<li>M. T. Toghani and G. I. Allen, “MP-Boost: Minipatch Boosting via Adaptive Feature and Observation Sampling”, In <em>IEEE International Conference on Big Data and Smart Computing (BigComp)</em>, 2021. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC8562710/">link</a></li>
<li>L. Gan and G. I. Allen, “Fast and Interpretable Consensus Clustering via Minipatch Learning”, <em>PLoS Computational Biology</em>, <strong>18</strong>:10, e1010577, 2022. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010577">link</a></li>
<li>L. Gan, L. Zheng, and G. I. Allen, “Model-Agnostic Confidence Intervals for Feature Importance: A Fast and Powerful Approach Using Minipatch Ensembles”, arXiv:2206.02088, 2022. <a href="https://arxiv.org/pdf/2206.02088">link</a></li>
<li>T. Yao, M. Wang, and G. I. Allen, “Fast and Accurate Graph Learning for Huge Data via Minipatch Ensembles”, arXiv:2110.12067, 2021. <a href="https://arxiv.org/pdf/2110.12067">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="quilting-patchwork-learning" class="level4">
<h4 class="anchored" data-anchor-id="quilting-patchwork-learning">Quilting &amp; Patchwork Learning</h4>
<div class="grid">
<div class="g-col-3">
<p><img src="images/quilting_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-9">
<p>In neuroscience, data integration, causal panel data and more, we often observe data in patches or blocks with huge portions of data that is not missing at random. We recently have developed new unsupervised approaches in this patchwork learning setting.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>G. Vinci, G. Dasarathy, and G. I. Allen, “Graph Quilting: Graphical Model Selection from Partially Observed Covariances”, arXiv:1912.05573, 2020. <a href="https://arxiv.org/pdf/1912.05573">link</a></li>
<li>A. Chang, L. Zheng and G. I. Allen, “Low-Rank Covariance Completion for Graph Quilting with Applications to Functional Connectivity”, arXiv:2209.08273, 2022. <a href="https://arxiv.org/pdf/2209.08273">link</a></li>
<li>L. Zheng and G. I. Allen, “Graphical Model Inference with Erosely Measured Data”, (To Appear) <em>Journal of American Statistical Association: Theory &amp; Methods</em>, arXiv:2210.11625, 2023. <a href="https://arxiv.org/pdf/2210.11625">link</a></li>
<li>A. Chang, L. Zheng, G. Dasarthy, G. I. Allen, ”Nonparanormal Graph Quilting with Applications to Calcium Imaging”, Stat, 12:1, e623, 2023. <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/sta4.623">link</a></li>
<li>L. Zheng, A. Chang, and G. I. Allen, “Cluster Quilting: Spectral Clustering for Patchwork Learning”, arXiv:2406.13833, 2024. <a href="https://arxiv.org/pdf/2406.13833">link</a></li>
</ul>
</details>
</div>
</div>
</section>
</section>
<section id="neuroscience" class="level3">
<h3 class="anchored" data-anchor-id="neuroscience">Neuroscience</h3>
<section id="connectomics" class="level4">
<h4 class="anchored" data-anchor-id="connectomics">Connectomics</h4>
<div class="grid">
<div class="g-col-5">
<p><img src="images/neuroscience-or-networks_thumbnail.jpg" class="img-fluid"></p>
</div>
<div class="g-col-7">
<p>Connectomics seeks to understand how brain regions or neurons are structurally and functionally connected. Our research has focused on two aspects: developing new techniques to discover patterns in connectomics data and developing new techniques to learn functional connections from neural activity.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>M. Narayan and G. I. Allen, “Mixed Effects Models for Resampled Network Statistics Improves Statistical Power to Find Differences in Multi-Subject Functional Connectivity” <em>Frontiers in Neuroscience</em>, <strong>10</strong>:108, 2016. <a href="https://www.frontiersin.org/journals/neuroscience/articles/10.3389/fnins.2016.00108/full">link</a></li>
<li>S. Tomson, M. Narayan, G. I. Allen, D. Eagleman, “Neural Networks of Synesthesia”, <em>Journal of Neuroscience</em>, <strong>33</strong>:35, 14098-14106, 2013. <a href="https://www.jneurosci.org/content/jneuro/33/35/14098.full.pdf">link</a></li>
<li>S. Tomson, M. Schreiner, M. Narayan, T. Rosser; N. Enrique, A. J. Silva, G. I. Allen, S. Y. Bookheimer, and C. Bearden, “Resting state functional MRI reveals abnormal network connectivity in Neurofibromatosis 1”, <em>Human Brain Mapping</em>, <strong>36</strong>:11, 4566-4581, 2015. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4619152/">link</a></li>
<li>M. Wang and G. I. Allen, “Thresholded Graphical Lasso Adjusts for Latent Variables”, <em>Biometrika</em>, <strong>110</strong>:3, 681-697, 2023. <a href="https://par.nsf.gov/servlets/purl/10470624">link</a></li>
<li>G. Vinci, G. Dasarathy, and G. I. Allen, “Graph Quilting: Graphical Model Selection from Partially Observed Covariances”, arXiv:1912.05573, 2020. <a href="https://arxiv.org/pdf/1912.05573">link</a></li>
<li>A. Chang, L. Zheng and G. I. Allen, “Low-Rank Covariance Completion for Graph Quilting with Applications to Functional Connectivity”, arXiv:2209.08273, 2022. <a href="https://arxiv.org/pdf/2209.08273">link</a></li>
<li>A. Chang and G. I. Allen, “Subbotin Graphical Models for Extreme Value Dependencies with Applications to Functional Neuronal Connectivity”, <em>Annals of Applied Statistics</em>, <strong>17</strong>:3, 2364-2386, 2023. <a href="https://arxiv.org/pdf/2106.11554">link</a></li>
<li>T. Yao, M. Wang, and G. I. Allen, “Fast and Accurate Graph Learning for Huge Data via Minipatch Ensembles”, arXiv:2110.12067, 2021. <a href="https://arxiv.org/pdf/2110.12067">link</a></li>
<li>A. Chang, T. Yao, and G. I. Allen, “Graphical Models and Dynamic Factor Models for Modeling Functional Brain Connectivity”, In <em>Proceedings of the IEEE Data Science Workshop</em>, 2019. <a href="https://ieeexplore.ieee.org/abstract/document/8755783?casa_token=QRdm3GqNkG4AAAAA:dWbhysYt6989yPF4GBEOMviZYfWV5CAbuBSveOiSV_FtUmaHYgbQ1hyaf_XmSizQGsoQHjbq">link</a></li>
<li>Z. Zhang, G. I. Allen, H. Zhu, D. Dunson, “Tensor Network Factorizations: Relationships between Human Brain Structural Connectomes and Traits”, <em>NeuroImage</em>, <strong>197</strong>:330-343, 2019. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6613218/">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="pattern-discovery" class="level4">
<h4 class="anchored" data-anchor-id="pattern-discovery">Pattern Discovery</h4>
<div class="grid">
<div class="g-col-5">
<p><img src="images/ecog_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-7">
<p>Our research seeks to discover scientifically interpretable patterns from large-scale neuroscience data using a combination of statistical and interpretable machine learning approaches. We have applied our techniques to many types of macro and micro-scale recording and imaging technologies including MRI, functional MRI, diffusion imaging, EEG, PET, ECoG, calcium imaging, spike trains, and many more.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>Z. Zhang, G. I. Allen, H. Zhu, D. Dunson, “Tensor Network Factorizations: Relationships between Human Brain Structural Connectomes and Traits”, <em>NeuroImage</em>, <strong>197</strong>:330-343, 2019. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6613218/">link</a></li>
<li>K. Geyer, F. Campbell, A. Chang, J. Magnotti, M. Beauchamp, G. I. Allen, “Interpretable Visualization and Higher-Order Dimension Reduction for ECoG Data”, In <em>Proceedings of the The International Workshop on Big Data Reduction</em> held with the <em>2020 IEEE International Conference on Big Data</em>, 2020. <a href="https://arxiv.org/pdf/2011.09447">link</a></li>
<li>J. Liu, L. Zheng, Z. Zhang, and G. I. Allen, “Joint Semi-Symmetric Tensor PCA for Integrating Multi-modal Populations of Networks”, arXiv:2312.14416, 2023. <a href="https://arxiv.org/pdf/2312.14416">link</a></li>
<li>G. I. Allen, and M. Weylandt, “Sparse and Functional Principal Components Analysis”, In <em>Proceedings of the IEEE Data Science Workshop</em>, 2019. <a href="https://arxiv.org/pdf/1309.2895">link</a></li>
<li>Y. Hu and G. I. Allen, “Local-Aggregate Modeling for Big-Data via Distributed Optimization: Applications to Neuroimaging”, <em>Biometrics</em>, <strong>71</strong>:4, 905-917, 2015. <a href="https://academic.oup.com/biometrics/article-abstract/71/4/905/7511581?redirectedFrom=PDF">link</a></li>
<li>. T. Yao, E. M. Sweeney, J. Nagorski, J. Shulman, and G. I. Allen, “Quantifying cognitive resilience in Alzheimer’s Disease: The Alzheimer’s Disease Cognitive Resilience Score”, PLoS One, 15:11, e0241707, 2020. <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0241707">link</a></li>
</ul>
</details>
</div>
</div>
</section>
</section>
<section id="ai-ethics" class="level3">
<h3 class="anchored" data-anchor-id="ai-ethics">AI Ethics</h3>
<section id="interpretable-machine-learning" class="level4">
<h4 class="anchored" data-anchor-id="interpretable-machine-learning">Interpretable Machine Learning</h4>
<div class="grid">
<div class="g-col-6">
<p><img src="images/iml_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-6">
<p>Recently, we have begun working on statistical challenges in interpretable machine learning. Our focus is on developing validation and uncertainty quantification strategies for machine learning interpretations including feature importance and unsupervised discoveries.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>G. I. Allen, L. Gan, L. Zheng, ”Interpretable Machine Learning for Discovery: Statistical Challenges and Opportunities”, Annual Review of Statistics and its Application, 11:97-121, 2024. <a href="https://www.annualreviews.org/content/journals/10.1146/annurev-statistics-040120-030919?crawler=true">link</a></li>
<li>L. Gan, L. Zheng, and G. I. Allen, “Model-Agnostic Confidence Intervals for Feature Importance: A Fast and Powerful Approach Using Minipatch Ensembles”, arXiv:2206.02088, 2022. <a href="https://arxiv.org/pdf/2206.02088">link</a></li>
</ul>
</details>
</div>
</div>
</section>
<section id="algorithmic-fairness" class="level4">
<h4 class="anchored" data-anchor-id="algorithmic-fairness">Algorithmic Fairness</h4>
<div class="grid">
<div class="g-col-5">
<p><img src="images/fairstacks_multi_constraint.png" class="img-fluid"></p>
</div>
<div class="g-col-7">
<p>Machine learning algorithms can often inadvertently discriminate against certain subgroups by exacerbating subtle biases in historic data. Our recent work in the area of algorithmic fairness has sought to develop ways to audit and interpret bias mitigation strategies in machine learning.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>C. O. Little, M. Weylandt, and G. I. Allen, “To the Fairness Frontier and Beyond: Identifying, Quantifying, and Optimizing the Fairness-Accuracy Pareto Frontier”, arXiv:2206.00074, 2022. <a href="https://arxiv.org/pdf/2206.00074">link</a></li>
<li>C. O. Little, D. H. Lina, and G. I. Allen, “Fair Feature Importance Scores for Interpreting Tree-Based Methods and Surrogates”, (To Appear), <em>Transactions on Machine Learning Research</em>, arXiv:2310.04352, 2024. <a href="https://openreview.net/pdf?id=72mDxlzRZ1">link</a></li>
<li>M. Navarro, C. O. Little, G. I. Allen, and S. Segarra, “Data Augmentation via Subgroup Mixup for Improving Fairness”, In <em>IEEE International Conference on Acoustics, Speech, and Signal Processing</em>, 7350-7354, 2024. <a href="https://arxiv.org/pdf/2309.07110">link</a></li>
</ul>
</details>
</div>
</div>
</section>
</section>
<section id="biomedicine" class="level3">
<h3 class="anchored" data-anchor-id="biomedicine">Biomedicine</h3>
<section id="genomics" class="level4">
<h4 class="anchored" data-anchor-id="genomics">Genomics</h4>
<div class="grid">
<div class="g-col-4">
<p><img src="images/fig_genomics_thumbnail.png" class="img-fluid"></p>
</div>
<div class="g-col-8">
<p>Much of our research program has been inspired by challenges in high-throughput genomics and multi-omics data. We have developed new statistical machine learning techniques and applied these to study genomic mechanisms in cancer and neurological diseases.</p>
<details>
<summary>
Key Publications:
</summary>
<ul>
<li>Y. W. Wan, G. I. Allen, and Z. Liu, “TCGA2STAT: Simple TCGA Data Access for Integrated Statistical Analysis in R”, <em>Bioinformatics</em>, <strong>32</strong>:6, 952-954, 2016. <a href="https://scholar.google.com/citations?view_op=view_citation&amp;hl=en&amp;user=gIUd12QAAAAJ&amp;cstart=20&amp;pagesize=80&amp;sortby=pubdate&amp;citation_for_view=gIUd12QAAAAJ:RHpTSmoSYBkC">link</a></li>
<li>G. I. Allen and Z. Liu, “A Local Poisson Graphical Model for Inferring Networks from Next Generation Sequencing Data”, <em>IEEE Transactions on NanoBioscience</em>, <strong>12</strong>:3, 1-10, 2013. <a href="https://ieeexplore.ieee.org/abstract/document/6578145">link</a></li>
<li>G. I. Allen and M. Maletic-Savatic, “Sparse Non-negative Generalized PCA with Applications to Metabolomics”, <em>Bioinformatics</em>, <strong>27</strong>:21, 3029-3035, 2011. <a href="https://academic.oup.com/bioinformatics/article/27/21/3029/218921">link</a></li>
<li>L. Gan and G. I. Allen, “Fast and Interpretable Consensus Clustering via Minipatch Learning”, <em>PLoS Computational Biology</em>, <strong>18</strong>:10, e1010577, 2022. <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010577">link</a></li>
<li>L. Gan, G. Vinci, and G. I. Allen, “Correlation Imputation for Single Cell RNA-seq”, <em>Journal of Computational Biology</em>, <strong>29</strong>:5, 465-482, 2022. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC9125575/">link</a></li>
</ul>
</details>
</div>
</div>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/genevera-allen\.com\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>